### PR TITLE
feat: Implement write access sharing

### LINKED ISSUE #376: Implement write access sharing

### PR DESCRIPTION
**Pull request by Oleksii-Klimov**

### CODE REVIEW

**Comment by ai-dial-actions** [comment]
Hey there and thank you for opening this pull request! 👋🏼

We require pull request titles to follow the [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/) and it looks like your proposed title needs to be adjusted.

Details:

```
No release type found in pull request title "Implement write access sharing". Add a prefix to indicate what kind of release this pull request corresponds to. For reference, see https://www.conventionalcommits.org/

Available types:
 - fix
 - hotfix
 - feat
 - feature
 - chore
```

File: src/main/java/com/epam/aidial/core/controller/AccessControlBaseController.java [hunk 1]
```diff
@@ -2,18 +2,22 @@
...
+import java.util.List;
+import java.util.Set;
+
 @AllArgsConstructor
 public abstract class AccessControlBaseController {
 
     final Proxy proxy;
     final ProxyContext context;
-    final boolean isWriteAccess;
+    final ResourceAccessType accessType;
```
Comment (astsiapanay) [inline src/main/java/com/epam/aidial/core/controller/AccessControlBaseController.java]: I think we can keep as is here and all rest of controllers. The permission can be either read or write.

File: src/main/java/com/epam/aidial/core/security/AccessService.java [hunk 2]
```diff
@@ -33,59 +47,53 @@ public AccessService(EncryptionService encryptionService,
...
-    public boolean hasWriteAccess(ResourceDescription resource, ProxyContext context) {
-        if (resource.isPublic()) {
-            return hasPublicAccess(List.of(resource), context, false);
+    public Set<ResourceAccessType> lookupPermissions(
+            ResourceDescription resource, ProxyContext context, Set<ResourceAccessType> toLookup) {
+        Set<ResourceAccessType> remainingAccess = new HashSet<>(toLookup);
+        Set<ResourceAccessType> result = new HashSet<>();
+        for (PermissionRule permissionRule : permissionRules) {
+            if (permissionRule.predicate.test(resource, context)) {
+                Set<ResourceAccessType> permissions = permissionRule.function.apply(resource, context);
```
Comment (astsiapanay) [inline src/main/java/com/epam/aidial/core/security/AccessService.java]: We can make implementation more simple with a bitmask: READ is the 1st bit, WRITE - the 2nd.
and later use OR on the bit mask and the function result
```
  bitmask |= res;
 if ( bitmask == (1 << ResourceAccessType.values.length) - 1)
   return toLookup;
```
Anyway it's up to you.

File: src/main/java/com/epam/aidial/core/controller/ShareController.java [hunk 3]
```diff
@@ -133,13 +130,13 @@ public Future<?> revokeSharedResources() {
         return context.getRequest()
                 .body()
                 .compose(buffer -> {
-                    ResourceLinkCollection request = getResourceLinkCollection(buffer, Operation.REVOKE);
+                    RevokeResourcesRequest request = getRevokeResourcesRequest(buffer, Operation.REVOKE);
```
Comment (astsiapanay) [inline src/main/java/com/epam/aidial/core/controller/ShareController.java:136]: We can keep as is. `RevokeResourcesRequest` is identical to the previous class str-re.

File: src/main/java/com/epam/aidial/core/data/RevokeResourcesRequest.java [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package com.epam.aidial.core.data;
+
+import lombok.Data;
+
+import java.util.Set;
+
+@Data
+public class RevokeResourcesRequest {
```
Comment (astsiapanay) [inline src/main/java/com/epam/aidial/core/data/RevokeResourcesRequest.java:11]: Do we really need this?

File: src/main/java/com/epam/aidial/core/data/SharedByMeDto.java [hunk 5]
```diff
@@ -1,26 +1,77 @@
...
+import static java.util.function.Predicate.not;
 
 @Data
-@AllArgsConstructor
-@NoArgsConstructor
 public class SharedByMeDto {
-    Map<String, Set<String>> resourceToUsers;
+    @JsonProperty("resourceToUsers")
+    Map<String, Set<String>> readableResourceToUsers;
+    Map<String, Map<String, Set<ResourceAccessType>>> resourcesWithPermissions;
```
Comment (astsiapanay) [inline src/main/java/com/epam/aidial/core/data/SharedByMeDto.java]: It looks ugly but I don't know a way how to extend the str-re and preserve backward compatibility.

File: src/main/java/com/epam/aidial/core/data/SharedResources.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+import java.util.EnumSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@Data
+public class SharedResources {
+    @JsonProperty("resources")
+    Set<ResourceLink> readableResources;
+    Map<String, Set<ResourceAccessType>> resourcesWithPermissions;
```
Comment (astsiapanay) [inline src/main/java/com/epam/aidial/core/data/SharedResources.java]: Can we replace `ResourceLink` with `SharedResource`?

File: src/main/java/com/epam/aidial/core/service/PublicationService.java [hunk 7]
```diff
@@ -363,7 +367,8 @@ private void prepareAndValidatePublicationRequest(ProxyContext context, String b
                 .toList();
 
         // validate if user has access to all target resources
-        boolean hasPublicAccess = accessService.hasPublicAccess(targetResources, context);
+        boolean hasPublicAccess = targetResources.stream()
```
Comment (astsiapanay) [inline src/main/java/com/epam/aidial/core/service/PublicationService.java]: Why did you change the access check from public access to has read access?

File: src/main/java/com/epam/aidial/core/controller/AccessControlBaseController.java (hunk 1 above)
Comment (Oleksii-Klimov) [reply src/main/java/com/epam/aidial/core/controller/AccessControlBaseController.java]: Reverted.

File: src/main/java/com/epam/aidial/core/controller/ShareController.java (hunk 3 above)
Comment (Oleksii-Klimov) [reply src/main/java/com/epam/aidial/core/controller/ShareController.java:136]: Updated. The revoke request can now remove specific permissions.

File: src/main/java/com/epam/aidial/core/data/RevokeResourcesRequest.java (hunk 4 above)
Comment (Oleksii-Klimov) [reply src/main/java/com/epam/aidial/core/data/RevokeResourcesRequest.java:11]: Implemented permission removal. Now it's needed.

File: src/main/java/com/epam/aidial/core/data/SharedResources.java (hunk 6 above)
Comment (Oleksii-Klimov) [reply src/main/java/com/epam/aidial/core/data/SharedResources.java]: Replaced.

File: src/main/java/com/epam/aidial/core/security/AccessService.java (hunk 2 above)
Comment (Oleksii-Klimov) [reply src/main/java/com/epam/aidial/core/security/AccessService.java]: I'll check if I can improve it.

File: src/main/java/com/epam/aidial/core/service/PublicationService.java (hunk 7 above)
Comment (Oleksii-Klimov) [reply src/main/java/com/epam/aidial/core/service/PublicationService.java]: Changed.

**Comment by Maxim-Gadalov** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-core-pr-377.nightly-test.deltixhub.io
>E2E tests status: success

**Comment by Oleksii-Klimov** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-core-pr-377.nightly-test.deltixhub.io
>E2E tests status: failed

File: src/main/java/com/epam/aidial/core/data/SharedResource.java [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+    public SharedResource withReadIfNoPermissions() {
+        return permissions == null || permissions.isEmpty()
+                ? withPermissions(EnumSet.of(ResourceAccessType.READ))
+                : this;
+    }
+
+    public SharedResource withAllIfNoPermissions() {
+        return permissions == null || permissions.isEmpty()
+                ? withPermissions(EnumSet.allOf(ResourceAccessType.class))
```
Comment (Maxim-Gadalov) [inline src/main/java/com/epam/aidial/core/data/SharedResource.java]: withpermissions(ResourceAccessType.ALL) ?

File: src/main/java/com/epam/aidial/core/service/PublicationService.java (hunk 7 above)
Comment (Maxim-Gadalov) [reply src/main/java/com/epam/aidial/core/service/PublicationService.java]: previous version was optimized, this one - is not, please use optimized version because permission check for public resources quite expensive

File: src/main/java/com/epam/aidial/core/security/AccessService.java [hunk 9]
```diff
@@ -33,116 +49,162 @@ public AccessService(EncryptionService encryptionService,
...
-     * @param resources - public resources
-     * @param context - context
-     * @param readOnly - true to check read only access, false to check write access as well
-     * @return true - if all provided resources are public and user has permissions to all of them, otherwise - false
-     */
-    private boolean hasPublicAccess(List<ResourceDescription> resources, ProxyContext context, boolean readOnly) {
-        boolean isAllPublic = resources.stream().allMatch(ResourceDescription::isPublic);
-        if (!isAllPublic) {
-            return false;
+            if (hasAdminAccess(context)) {
```
Comment (Maxim-Gadalov) [inline src/main/java/com/epam/aidial/core/security/AccessService.java]: hasAdminAccess can be evaluated outside of the loop

File: src/main/java/com/epam/aidial/core/service/RuleService.java [hunk 10]
```diff
@@ -82,31 +81,13 @@ public Map<String, List<Rule>> listRules(ResourceDescription resource) {
...
-        boolean hasAccess = false;
-        for (ResourceDescription resource : resources) {
-            hasAccess = evaluate(context, resource, rules, cache);
-            if (!hasAccess) {
-                return false;
-            }
-        }
-
-        return hasAccess;
+        return evaluate(context, resource, rules, cache);
```
Comment (Maxim-Gadalov) [inline src/main/java/com/epam/aidial/core/service/RuleService.java]: why you removed an optimization?

File: src/main/java/com/epam/aidial/core/security/AccessService.java [hunk 11]
```diff
@@ -33,116 +49,162 @@ public AccessService(EncryptionService encryptionService,
...
+            Set<ResourceDescription> resources, ProxyContext context) {
+        Map<ResourceDescription, Set<ResourceAccessType>> result = new HashMap<>();
+        for (ResourceDescription resource : resources) {
+            if (!PublicationService.isReviewResource(resource)) {
+                continue;
+            }
 
-        if (hasAdminAccess(context)) {
-            return true;
+            if (hasAdminAccess(context)) {
```
Comment (Maxim-Gadalov) [inline src/main/java/com/epam/aidial/core/security/AccessService.java]: same here

File: src/main/java/com/epam/aidial/core/security/AccessService.java [hunk 12]
```diff
@@ -11,16 +12,31 @@
...
+import java.util.function.BiFunction;
+
 
 public class AccessService {
 
     private final EncryptionService encryptionService;
     private final ShareService shareService;
     private final RuleService ruleService;
     private final List<Rule> adminRules;
+    private final List<PermissionRule> permissionRules = List.of(
```
Comment (Maxim-Gadalov) [inline src/main/java/com/epam/aidial/core/security/AccessService.java:35]: Nit: most of the use-cases related to chat-completion and own resources. I would rearrange checks to the following:
checkOwnResource - cheep
checkAutoShared - cheep
checkAppResourceAccess - cheep
checkSharedResource - mid
checkPublishedResource - expensive
checkReviewResource - cheep (rare case)

File: src/main/java/com/epam/aidial/core/service/InvitationService.java [hunk 13]
```diff
@@ -126,44 +128,58 @@ public InvitationCollection getMyInvitations(String bucket, String location) {
         return new InvitationCollection(new HashSet<>(invitationMap.getInvitations().values()));
     }
 
-    public void cleanUpResourceLinks(String bucket, String location, Set<ResourceLink> resourcesToCleanUp) {
+    public void cleanUpResourceLinks(
+            String bucket, String location, Map<String, Set<ResourceAccessType>> permissionsToCleanUp) {
```
Comment (Maxim-Gadalov) [inline src/main/java/com/epam/aidial/core/service/InvitationService.java]: you are cleaning up resources not permissions

File: src/main/java/com/epam/aidial/core/service/InvitationService.java [hunk 14]
```diff
@@ -126,44 +128,58 @@ public InvitationCollection getMyInvitations(String bucket, String location) {
...
         ResourceDescription resource = ResourceDescription.fromDecoded(ResourceType.INVITATION, bucket, location, INVITATION_RESOURCE_FILENAME);
         resourceService.computeResource(resource, state -> {
             InvitationsMap invitations = ProxyUtil.convertToObject(state, InvitationsMap.class);
             if (invitations == null) {
                 return null;
             }
             Map<String, Invitation> invitationMap = invitations.getInvitations();
-            List<Invitation> invitationsToRemove = new ArrayList<>();
+            List<String> invitationsToRemove = new ArrayList<>();
```
Comment (Maxim-Gadalov) [inline src/main/java/com/epam/aidial/core/service/InvitationService.java:143]: previous version removes resource links from invitation, this version only removes invitations completely if links remain. Please add a test for that case

File: src/main/java/com/epam/aidial/core/service/InvitationService.java [hunk 15]
```diff
@@ -126,44 +128,58 @@ public InvitationCollection getMyInvitations(String bucket, String location) {
...
             InvitationsMap invitations = ProxyUtil.convertToObject(state, InvitationsMap.class);
             if (invitations == null) {
                 return null;
             }
             Map<String, Invitation> invitationMap = invitations.getInvitations();
             for (Invitation invitation : invitationMap.values()) {
-                Set<ResourceLink> invitationResourceLinks = invitation.getResources();
-                if (invitationResourceLinks.remove(sourceLink)) {
-                    invitationResourceLinks.add(destinationLink);
+                Set<SharedResource> invitationResourceLinks = invitation.getResources();
```
Comment (Maxim-Gadalov) [inline src/main/java/com/epam/aidial/core/service/InvitationService.java]: why do we need to overcomplecate? resources in the invitation stored in set, if somebody put same resource with different permissions we need to return a 400.

File: src/main/java/com/epam/aidial/core/service/ShareService.java [hunk 16]
```diff
@@ -144,134 +164,178 @@ public void acceptSharedResources(String bucket, String location, String invitat
...
                     }
 
                     return ProxyUtil.convertToString(sharedByMeDto);
                 });
             }
-        }
+        });
     }
 
     public void discardSharedAccess(String bucket, String location, ResourceLinkCollection request) {
```
Comment (Maxim-Gadalov) [inline src/main/java/com/epam/aidial/core/service/ShareService.java:310]: I'd remove this class and use a new one ignoring permissions field

File: src/main/java/com/epam/aidial/core/data/SharedByMeDto.java [hunk 17]
```diff
@@ -1,26 +1,73 @@
...
 import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
 
 @Data
-@AllArgsConstructor
-@NoArgsConstructor
 public class SharedByMeDto {
-    Map<String, Set<String>> resourceToUsers;
+    @JsonProperty("resourceToUsers")
```
Comment (Maxim-Gadalov) [inline src/main/java/com/epam/aidial/core/data/SharedByMeDto.java:20]: I don't understand what's going on here, you merged an old state to the new and deal with both anyway

File: src/main/java/com/epam/aidial/core/service/ShareService.java [hunk 18]
```diff
@@ -144,134 +164,178 @@ public void acceptSharedResources(String bucket, String location, String invitat
...
 
                 return ProxyUtil.convertToString(dto);
             });
 
             ResourceDescription sharedWithMe = getShareResource(ResourceType.SHARED_WITH_ME, resourceType, bucket, location);
             resourceService.computeResource(sharedWithMe, state -> {
-                ResourceLinkCollection collection = ProxyUtil.convertToObject(state, ResourceLinkCollection.class);
-                if (collection == null) {
-                    collection = new ResourceLinkCollection(new HashSet<>());
+                SharedResources sharedResources = ProxyUtil.convertToObject(state, SharedResources.class);
```
Comment (Maxim-Gadalov) [inline src/main/java/com/epam/aidial/core/service/ShareService.java:189]: is it possible to simplify this code or move it to the separate method?

File: src/main/java/com/epam/aidial/core/controller/FileMetadataController.java [hunk 19]
```diff
@@ -33,9 +37,10 @@ protected Future<?> handle(ResourceDescription resource) {
         if (limit < 0 || limit > 1000) {
             return context.respond(HttpStatus.BAD_REQUEST, "Limit is out of allowed range: [0, 1000]");
         }
+        PermissionsFetcher permissionsFetcher = PermissionsFetcher.of(context, accessService);
         return proxy.getVertx().executeBlocking(() -> {
             try {
-                MetadataBase metadata = storage.listMetadata(resource, token, limit, recursive);
+                MetadataBase metadata = storage.listMetadata(resource, permissionsFetcher, token, limit, recursive);
```
Comment (Maxim-Gadalov) [inline src/main/java/com/epam/aidial/core/controller/FileMetadataController.java]: I would separate fetching permissions from metadata listing. let's keep abstractions simple

File: src/main/java/com/epam/aidial/core/controller/PublicationController.java [hunk 20]
```diff
@@ -161,7 +162,9 @@ public Future<?> listPublishedResources() {
                     ListPublishedResourcesRequest request = ProxyUtil.convertToObject(body, ListPublishedResourcesRequest.class);
                     String bucketLocation = BlobStorageUtil.buildInitiatorBucket(context);
                     String bucket = encryptService.encrypt(bucketLocation);
-                    return vertx.executeBlocking(() -> publicationService.listPublishedResources(request, bucket, bucketLocation), false);
+                    PermissionsFetcher permissionsFetcher = PermissionsFetcher.of(context, accessService);
+                    return vertx.executeBlocking(() -> publicationService.listPublishedResources(
```
Comment (Maxim-Gadalov) [inline src/main/java/com/epam/aidial/core/controller/PublicationController.java]: same here, we can adjust metadata response with permissions here

File: src/main/java/com/epam/aidial/core/service/PermissionsFetcher.java [hunk 21]
```diff
@@ -0,0 +1,22 @@
...
+import java.util.Set;
+
+@FunctionalInterface
+public interface PermissionsFetcher {
+    PermissionsFetcher EMPTY = resources -> Map.of();
+
+    Map<ResourceDescription, Set<ResourceAccessType>> fetch(Set<ResourceDescription> resources);
+
+    static PermissionsFetcher of(ProxyContext context, AccessService accessService) {
+        return Boolean.parseBoolean(context.getRequest().getParam("permissions", "false"))
```
Comment (Maxim-Gadalov) [inline src/main/java/com/epam/aidial/core/service/PermissionsFetcher.java]: I think we need to keep all request parameters in controllers and provide a generic service, any service may later use this functionality

File: src/main/java/com/epam/aidial/core/data/SharedResource.java (hunk 8 above)
Comment (Oleksii-Klimov) [reply src/main/java/com/epam/aidial/core/data/SharedResource.java]: I'd like to maintain mutability for consistency, as there are many places where these objects are expected to be mutable. I have replaced it with the following for clarity:
```java
EnumSet.copyOf(ResourceAccessType.ALL)
```

File: src/main/java/com/epam/aidial/core/security/AccessService.java (hunk 9 above)
Comment (Oleksii-Klimov) [reply src/main/java/com/epam/aidial/core/security/AccessService.java]: Extracted into a separate rule.

File: src/main/java/com/epam/aidial/core/service/RuleService.java (hunk 10 above)
Comment (Oleksii-Klimov) [reply src/main/java/com/epam/aidial/core/service/RuleService.java]: It wasn't intentional. Reverted.

File: src/main/java/com/epam/aidial/core/security/AccessService.java (hunk 11 above)
Comment (Oleksii-Klimov) [reply src/main/java/com/epam/aidial/core/security/AccessService.java]: Done.

File: src/main/java/com/epam/aidial/core/security/AccessService.java (hunk 12 above)
Comment (Oleksii-Klimov) [reply src/main/java/com/epam/aidial/core/security/AccessService.java:35]: Done.

File: src/main/java/com/epam/aidial/core/service/InvitationService.java (hunk 13 above)
Comment (Oleksii-Klimov) [reply src/main/java/com/epam/aidial/core/service/InvitationService.java]: Renamed and slightly refactored.

File: src/main/java/com/epam/aidial/core/service/InvitationService.java (hunk 14 above)
Comment (Oleksii-Klimov) [reply src/main/java/com/epam/aidial/core/service/InvitationService.java:143]: Fixed.

File: src/main/java/com/epam/aidial/core/service/PublicationService.java (hunk 7 above)
Comment (Oleksii-Klimov) [reply src/main/java/com/epam/aidial/core/service/PublicationService.java]: Done.

File: src/main/java/com/epam/aidial/core/data/SharedByMeDto.java (hunk 17 above)
Comment (Oleksii-Klimov) [reply src/main/java/com/epam/aidial/core/data/SharedByMeDto.java:20]: Refactored as discussed.

File: src/main/java/com/epam/aidial/core/service/ShareService.java (hunk 18 above)
Comment (Oleksii-Klimov) [reply src/main/java/com/epam/aidial/core/service/ShareService.java:189]: Done.

File: src/main/java/com/epam/aidial/core/controller/FileMetadataController.java (hunk 19 above)
Comment (Oleksii-Klimov) [reply src/main/java/com/epam/aidial/core/controller/FileMetadataController.java]: Done.

File: src/main/java/com/epam/aidial/core/controller/PublicationController.java (hunk 20 above)
Comment (Oleksii-Klimov) [reply src/main/java/com/epam/aidial/core/controller/PublicationController.java]: Done.

File: src/main/java/com/epam/aidial/core/service/PermissionsFetcher.java (hunk 21 above)
Comment (Oleksii-Klimov) [reply src/main/java/com/epam/aidial/core/service/PermissionsFetcher.java]: Done.

File: src/main/java/com/epam/aidial/core/service/InvitationService.java (hunk 15 above)
Comment (Oleksii-Klimov) [reply src/main/java/com/epam/aidial/core/service/InvitationService.java]: The structure doesn't imply that it's forbidden. I think it's fine to merge permissions on our side.

File: src/main/java/com/epam/aidial/core/ProxyContext.java [hunk 22]
```diff
@@ -134,4 +134,8 @@ public String getProject() {
     public List<String> getExecutionPath() {
         return proxyApiKeyData == null ? null : proxyApiKeyData.getExecutionPath();
     }
+
+    public boolean getBooleanParam(String name) {
```
Comment (Maxim-Gadalov) [inline src/main/java/com/epam/aidial/core/ProxyContext.java]: it's not clear from method name what boolean param you are returning. also it's good to mention default behavior in java doc

File: src/main/java/com/epam/aidial/core/data/SharedByMeDto.java [hunk 23]
```diff
@@ -1,26 +1,111 @@
...
+    }
+
+    public void removePermissionsFromResource(String url, Set<ResourceAccessType> permissionsToRemove) {
+        for (ResourceAccessType permission : permissionsToRemove) {
+            Map<String, Set<String>> usersMap = getUserMapForPermission(permission);
+            usersMap.remove(url);
+        }
+    }
+
+    public Map<String, Set<ResourceAccessType>> getAggregatedPermissions() {
```
Comment (Maxim-Gadalov) [inline src/main/java/com/epam/aidial/core/data/SharedByMeDto.java:81]: please, make sure that all get methods are not serialized in storage

File: src/main/java/com/epam/aidial/core/service/ShareService.java (hunk 16 above)
Comment (Oleksii-Klimov) [reply src/main/java/com/epam/aidial/core/service/ShareService.java:310]: I think it may be misleading.

File: src/main/java/com/epam/aidial/core/service/RuleService.java [hunk 24]
```diff
@@ -157,8 +152,7 @@ private static boolean evaluate(ProxyContext context,
             return true;
         }
 
-        String folderUrl = ruleUrl(resource);
-        Boolean evaluated = cache.get(folderUrl);
+        Boolean evaluated = cache.get(resource);
```
Comment (Maxim-Gadalov) [inline src/main/java/com/epam/aidial/core/service/RuleService.java]: previous version of a cache was more optimized.
ResourceDescription contain resource type and path
however ruleUrl removes resourceType

this means if you have file and conversation in same rule context ("public/folder1")
your implementation will miss the cache

File: src/main/java/com/epam/aidial/core/service/RuleService.java [hunk 25]
```diff
@@ -167,11 +161,14 @@ private static boolean evaluate(ProxyContext context,
...
 
         if (evaluated) {
-            List<Rule> folderRules = rules.get(folderUrl);
+            List<Rule> folderRules = rules.get(ruleUrl(resource));
             evaluated = folderRules == null || RuleMatcher.match(context, folderRules);
         }
 
-        cache.put(folderUrl, evaluated);
+        cache.put(resource, evaluated);
+        if (originalResource != resource) {
```
Comment (Maxim-Gadalov) [inline src/main/java/com/epam/aidial/core/service/RuleService.java]: can you explain why we need this?

File: src/main/java/com/epam/aidial/core/service/ShareService.java [hunk 26]
```diff
@@ -108,22 +116,22 @@ public SharedResourcesResponse listSharedByMe(String bucket, String location, Li
...
-        if (resourceLinks.isEmpty()) {
+        Set<SharedResource> sharedResources = request.getResources();
+        if (sharedResources.isEmpty()) {
             throw new IllegalArgumentException("No resources provided");
         }
 
-        Set<ResourceLink> normalizedResourceLinks = new HashSet<>();
-        for (ResourceLink resourceLink : resourceLinks) {
-            String url = resourceLink.url();
+        List<SharedResource> normalizedResourceLinks = new ArrayList<>(sharedResources.size());
```
Comment (Maxim-Gadalov) [inline src/main/java/com/epam/aidial/core/service/ShareService.java:125]: after normalization
folder@and folder%40 may point to the same resource
why you removed set here?

File: src/test/java/com/epam/aidial/core/ResourceBaseTest.java [hunk 27]
```diff
@@ -211,12 +213,35 @@ static void verify(Response response, int status, String body) {
     static void verifyJson(Response response, int status, String body) {
         assertEquals(status, response.status());
         try {
-            assertEquals(ProxyUtil.MAPPER.readTree(body).toPrettyString(), ProxyUtil.MAPPER.readTree(response.body()).toPrettyString());
+            JsonNode expected = sortPermissions(ProxyUtil.MAPPER.readTree(body));
+            JsonNode actual = sortPermissions(ProxyUtil.MAPPER.readTree(response.body()));
```
Comment (Maxim-Gadalov) [inline src/test/java/com/epam/aidial/core/ResourceBaseTest.java]: sorting permissions 3 times?

File: src/test/java/com/epam/aidial/core/ResourceBaseTest.java [hunk 28]
```diff
@@ -211,12 +213,35 @@ static void verify(Response response, int status, String body) {
...
         }
     }
 
+    static void verifyMetadata(MetadataBase expected, String actual) throws JsonProcessingException {
+        assertEquals(
+                sortPermissions(ProxyUtil.MAPPER.readTree(ProxyUtil.MAPPER.writeValueAsString(expected))),
+                sortPermissions(ProxyUtil.MAPPER.readTree(actual)));
+    }
+
+    private static JsonNode sortPermissions(JsonNode root) {
```
Comment (Maxim-Gadalov) [inline src/test/java/com/epam/aidial/core/ResourceBaseTest.java]: this method is not generic, I would add permissions in a right order or think about generic way to deal with set-like structures

File: src/test/java/com/epam/aidial/core/ShareApiTest.java [hunk 29]
```diff
@@ -120,7 +121,145 @@ public void testShareWorkflow() {
...
+        verifyJson(response, 200, """
+                {
+                  "resources": []
+                }
+                """);
+    }
+
+    @Test
+    public void testShareForWrite() {
+        // check no conversations shared with me
```
Comment (Maxim-Gadalov) [inline src/test/java/com/epam/aidial/core/ShareApiTest.java:146]: mentioned conversations, but files used

**Comment by Maxim-Gadalov** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-core-pr-377.nightly-test.deltixhub.io
>E2E tests status: failed

File: src/main/java/com/epam/aidial/core/data/SharedByMeDto.java (hunk 23 above)
Comment (Oleksii-Klimov) [reply src/main/java/com/epam/aidial/core/data/SharedByMeDto.java:81]: Ignoring.

File: src/main/java/com/epam/aidial/core/service/RuleService.java (hunk 24 above)
Comment (Oleksii-Klimov) [reply src/main/java/com/epam/aidial/core/service/RuleService.java]: Fixed.

File: src/main/java/com/epam/aidial/core/service/RuleService.java (hunk 25 above)
Comment (Oleksii-Klimov) [reply src/main/java/com/epam/aidial/core/service/RuleService.java]: Removed.

File: src/main/java/com/epam/aidial/core/service/ShareService.java (hunk 26 above)
Comment (Oleksii-Klimov) [reply src/main/java/com/epam/aidial/core/service/ShareService.java:125]: Checking for duplicates now.

File: src/test/java/com/epam/aidial/core/ResourceBaseTest.java (hunk 28 above)
Comment (Oleksii-Klimov) [reply src/test/java/com/epam/aidial/core/ResourceBaseTest.java]: In JSON, either sets or lists are represented identically. The order of elements in sets is usually not guaranteed (unless it's a special type of set) and may change from run to run.

File: src/test/java/com/epam/aidial/core/ShareApiTest.java (hunk 29 above)
Comment (Oleksii-Klimov) [reply src/test/java/com/epam/aidial/core/ShareApiTest.java:146]: Fixed.

**Comment by Maxim-Gadalov** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-core-pr-377.nightly-test.deltixhub.io
>E2E tests status: failed

**Comment by Maxim-Gadalov** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-core-pr-377.nightly-test.deltixhub.io
>E2E tests status: failed

**Comment by Oleksii-Klimov** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-core-pr-377.nightly-test.deltixhub.io
>E2E tests status: success

**Review by Maxim-Gadalov — APPROVED** [review]
